# general requirements


These practicals are built around [R](https://rstudio-education.github.io/hopr/starting.html).

They were tested on R version 4.0.2, but they should work on any versio, as I only used the more basic dependencies (no ggstuff or the tidyverse ...)


```
install.packages("BiocManager")
BiocManager::install("msa")
install.packages("phangorn")
install.packages("ape")
install.packages("phytools")
install.packages("seqinr")
install.packages("apex")
install.packages("ips")
install.packages("Biostrings")
```


---


# dataset building:

In the beginning we need to choose the group whose phylogeny will be inferred and the locus we will use for inferring their phylogeny. 
On [NCBI](https://www.ncbi.nlm.nih.gov/) select “Nucleotide” from the drop-down menu at the top and use the group and the gene names as queries, such as

- Mantodea AND Histone H3 
- Candovia AND cytochrome oxidase subunit I 
- txid70920[Organism:exp] AND small subunit ribosomal RNA gene 

For a set of fifteen/twenty species of the group of interest you should download a protein coding gene gene.
When you decide to download a given sequence, you have to place it in a [FASTA-formatted](https://en.wikipedia.org/wiki/FASTA_format) file.
Finally, you have to select outgroup sequences. You can find an example of a starting file [here](https://github.com/for-giobbe/Rphy/blob/main/sequences/COX1.nt.fa).