# Multiple Sequence Alignment & Filtering

Libraries:
- [Biostrings](https://kasperdanielhansen.github.io/genbioconductor/html/Biostrings.html#overview)
- [msa]
- [ape](https://emmanuelparadis.github.io/)

File types:
- DNAStringSet - from Biostring
- DNAbin - from ape 

---


## intro: 

Before starting this practical set the appropriate working directory and load the necessary libraries in R:

```
setwd("~/.../RPHY/")
library(Biostrings) 
library("msa")
library("ape")
```

All phylogenetic methods require a set of homologous characters! The first step is to infer which nucleotides / codons / aminocid residues are homologous to each other, 
so that differences among these nucleotides result only from changes that convey descent information. 
This process is called Multiple Sequence Alignment (MSA) and is often followed by an additional step of detection and exclusion of those sequences whose homology we are uncertain of.
Due to the large amount of data which we process nowadays, this step is often overlooked and as a result it's quite easy to find misaligned loci in modern datasets. 
Nonetheless this is a key step, which may affect downstream analyses (see further reading for more).

In this tutorial, we will use the most popular tools for multiple sequence alignment:

- Muscle [(Edgar 2004)](https://doi.org/10.1093/nar/gkh340)

- Clustal [(Higgins and Sharp 1988)](https://doi.org/10.1016/0378-1119(88)90330-7)

- T-Coffee [(Notredame et al., 2000)](https://doi.org/10.1006/jmbi.2000.4042)

And subsequently filter the alignments using:

- TrimAL [(Talavera & Castresana, 2007)](https://academic.oup.com/sysbio/article/56/4/564/1682121)

- a custom approach based on branch lengths!


---


As you can see, in the [fasta format](https://en.wikipedia.org/wiki/FASTA_format) each record consists of an ID and a sequence, of which the ID is always on a single line that starts with the ">" symbol, followed by lines containing the sequence. 
The sequences are not aligned yet: this is the reason why they contain no gaps and differ in length.
The use of short and simple IDs is strongly recommended because many programs or scripts may not work if you use spaces or hyphens.

At this stage you should have assembled your dataset into a multi-fasta file, which should look like [this](https://raw.githubusercontent.com/for-giobbe/Rphy/main/sequences/COX1.nt.fa). 
It's very important that each one of you has it's own dastaset as this may provide natural examples of problems in analyses, 
but If you still do not have a dataset use [the one I am using](https://github.com/for-giobbe/Rphy/main/sequences/COX1.nt.fa).

Let's start by importing the sequences in R:

```
sequences <- readDNAStringSet("sequences/COX1.nt.fa")
str(sequences)
```


---


## aligning with Muscle:

Muscle is a very fast and versatile tool. Let's start by typing:

```
alignment_muscle2 <- msa(sequences, "Muscle", gapOpening=0, gapExtension=0)
```

Muscle contains many parameters, which you are encouraged to explore: important ones are the gap-opening and gap-extension penalty which by default are 400 and 0. 
Let's try a different value, using the code:

```
alignment_muscle2 <- msa(sequences, "Muscle", gapOpening=0, gapExtension=0
```

We can then inspect the outcome there is any difference between the two gap-opening penalty values and using this code:

```
checkAlignment(alignment_muscle1, plot = F)
```

and

```
checkAlignment(alignment_muscle2, plot = F)
```

Instead for a visual representation:

```
par(mfcol = c(2, 1))
image(alignment_muscle1, show.labels = F)
image(alignment_muscle2, show.labels = F)
```

![alt text](https://github.com/for-giobbe/Rphy/blob/main/figures/Figure_1.1.png)


## aligning with Clustal:

```
alignment_clustalO <- msa(sequences, "ClustalOmega")
```

Sometimes aligners which are not aware of the underlying structure of codons in PCGs can introduce some artificial stop codons, which are more often than not alignment errors.
Insertions and/or deletions ("indels") with lengths that are not multiples of three will disrupt the reading frame and cause large changes to the protein structure which are usually strongly selected against and rarely found in alignments of protein-coding sequences.
In cases where the placement of indels is ambiguous, information about the reading frame can be used to optimize the positioning of these indels, 
by translating the PCG from nucleotides into aminoacids, align the aminoacids and then retrotranslate the aminoacids alignment to nucleotides. 

Also it's important to notice that one can carry out phylogenetic inference on nucleotides and/or aminoacids: the first ones convey more information (due to the genetic code degeneracy) but that information can sometimes non suitable for inferring phylogenies (i.e. substitution saturation).

We change from to
```
alignment_clustal <- msa(sequences, "ClustalOmega")
alignment_clustal <- as.DNAbin(alignment_clustal)
par(mfcol = c(1, 3))
checkAlignment(alignment_clustal, what = 2:4)
checkAlignment(alignment_clustal, what = 3:4)
checkAlignment(alignment_clustal, what = 4:4)
```

![alt text](https://github.com/for-giobbe/Rphy/blob/main/figures/Figure_1.2.png)

Next we are moving on-line: T-Coffe is a popular aligner with an incredibly nice [server](http://tcoffee.crg.cat/) which I strongly encourage to explore. But today we are focusing on just two approaches.


---


## One aligner to rule them all: [M-coffe](https://tcoffee.crg.eu/apps/tcoffee/do:mcoffee).


The idea behind M-COFFE is to combine many of the more popular aligners - including Clustal and Muscle!
We can give it a try using the default combination of aligners (Mkalign, Muscle & MAFFT) and then take a look at the outputs:


* the ```.aln``` is the alignment file in the clustal format.
* the ```.dnd``` file is the guide tree generated in the process.
* the ```.html``` file mirrors the .aln file, it's just easier to open by double clicking on it.

As you can see M-Coffee is combining multiple aligners into one and evaluating the uncertainty between their outcome. 


---


## The tradeoff between speed and accuracy: PSI-Coffe.

The more accurate method (IMHO) is to rely on. This approach is quite useful when dealing with distantly related proteins and can be a game-changer when accurate alignments are necessary, for example for inferences on selection regimes.
We can easily use this method with the string:


---


## Importing alignments.


```
alignement <- read.alignment("Accipitriformes_db.fa", format = "fasta")
alignement <- as.DNAbin(alignment_muscle)
alview(alignement, file = "", uppercase = TRUE, showpos = TRUE)

par(mfcol = c(2, 2))

for (x in c("a", "g", "c", "t"))
  image(alignement, x, "black", cex.lab = 0.5, cex.axis = 0.7)
  par(mfcol = c(1, 1))
```


![alt text](https://github.com/for-giobbe/Rphy/blob/main/figures/Figure_1.3.png)


---


## alignment filtering:

The quality of multiple sequence alignments plays an important role in the accuracy of phylogenetic inference. It has been shown that removing certain positions (generally the ones which show an ambiguous homology and/or are highly variable) can improve the overall performance of many phylogenetic reconstruction methods. 
While dealing modern phylogenetic dataset, which consists of hundred to thousands of alignments is not possible to have a manual curation for each one, and thus it is necessary to automatically remove alignment errors.

In this tutorial we will use Gblocks:  this software will select blocks of conserved sites, which can be defined with many custom parameters, described in the [manual](http://molevol.cmima.csic.es/castresana/Gblocks/Gblocks_documentation.html)

We will use the susequent parameters and leave the rest to the default:

* -t= p (protein) d (DNA) c (codons) 
* -b4 Minimum Length Of A Block
* -b5 Allowed Gap Positions

Let's use this string one PCGs gene:

```
Gblocks ND2_p_aligned.n.fasta -t=d -b4=5 -b5=a
```

and this for ncRNAs:

```
Gblocks 12S_xinsi_op7_aligned.fasta -t=d -b4=5 -b5=a
```

The analyses will generate two kind of files:

* ```-gb```: filtered fasta files 
* ```-gb.htm```: for visualising which position got removed

We have to slightly reformat the Gblocks outputs, using a combination of ```tr``` and ```sed```:

```cat ND2_p_aligned.n.fasta-gb | tr -d " " | sed 's/_1//g'  > ND2_p_aligned.n.gb.fasta```

and

```cat 12S_xinsi_op7_aligned.fasta-gb | tr -d " " | sed 's/_1//g'  > 12S_xinsi_op7_aligned.gb.fasta```












---


## alignment filtering pt. 2:

```
dm  <- dist.ml(alignment_clustal)
treeUPGMA  <- upgma(dm)
treeNJ  <- NJ(dm)

for (x in 1:length(treeNJ$tip.label)) {
  if (treeNJ$edge.length[x] > (sum(treeNJ$edge.length) * 0.1)) {
    print(treeNJ$tip.label[x])
  }
}
```


















Many other alternatives are possible, and all this tools can be extremely helpfull in removing noise and ameliorating certain characteristic of phylogenetic datasets which can affect subsequent inferences, such as substitution saturation & compositional heterogeneity. 
Here are the most popular:

[Aliscore](https://www.zfmk.de/de/forschung/forschungszentren-und-gruppen/aliscore) [(KÃ¼ck et al., 2014)](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-15-294)

[BMGE](https://research.pasteur.fr/en/software/bmge-block-mapping-and-gathering-with-entropy/) [(Criscuolo & Gibaldo)](https://bmcevolbiol.biomedcentral.com/articles/10.1186/1471-2148-10-210)

Other approaches for filtering are possible and useful when dealing with a large number of loci. Positions, genes and entire alignment can be filtered out on several crietria, including:

* length
* occupancy
* dNdS
* ...

Consider also that a very stringent filtering can be a terrible approach in certain scenarios, for example when the phylogenetic signal and/or the amount of positions considered are scarce.

---


## conclusion

In the end you should have an aligned & filtered MSA, named something like [12S_xinsi_op7_aligned.gb.fasta](https://github.com/for-giobbe/phy/blob/master/examples/12S_xinsi_op7_aligned.gb.fasta) and [ND2_p_aligned.n.gb.fasta](https://github.com/for-giobbe/phy/blob/master/examples/ND2_p_aligned.n.gb.fasta).
An underlying theme of bioinformatics should start to become clear: every file should have a name which should convey most, if not all, the information regarding it.


---


## further reading: 

[very interesting paper on multiple sequence alignment uncertainty pt.1 ](https://academic.oup.com/gbe/article/7/8/2102/556628).

[very interesting paper on multiple sequence alignment uncertainty pt.2](https://science.sciencemag.org/content/319/5862/473?casa_token=t07ptffISm4AAAAA:j5l4US_y_GHOMduYw6R-MhyM7YUpa__08lw45l455DAU3tGFNKYlV40ZH0Si5w48Xl1gTEqsocLVvaE).