# Multiple Sequence Alignment & Filtering


Libraries:
- [Biostrings](https://kasperdanielhansen.github.io/genbioconductor/html/Biostrings.html#overview)
- [phangorn](https://klausvigo.github.io/phangorn/)
- [msa](https://bioconductor.org/packages/devel/bioc/vignettes/msa/inst/doc/msa.pdf)
- [ape](https://emmanuelparadis.github.io/)
- [seqinr](https://www.rdocumentation.org/packages/seqinr/versions/4.2-36)

File types:
- DNAStringSet - from Biostring
- DNAbin - from ape 
- alignment" - from seqinr
- phyDat - from phangorn

---


## intro: 


Before starting this practical set the appropriate working directory and load the necessary libraries in R:


```
setwd("~/.../RPHY/")
library(Biostrings) 
library("msa")
library("ape")
library("seqinr")
library("phangorn")
```


All phylogenetic methods require a set of homologous characters! The first step is to infer which nucleotides / codons / ... / aminocids are homologous to each other, 
so that differences among the charachters considered in the analyses results only from changes that convey descent information. 
After having attained a Multiple Sequence Alignment (MSA), additional steps of exclusion of sequences of uncertain homology is often performed.
Due to the large amount of data which we process nowadays, this step is often overlooked and as a result it's quite easy to find misaligned loci in modern datasets. 
Nonetheless this is a key step, which may affect downstream analyses (see the "further reading" section for more on this aspect).

Let's start by "simulating" and aligning some sequences to instantly get a grasp of what it's all about:

```
nt_seq <- c("ATATCGGTA", "ACTATATATGCTG", "AGTGACAGGGTT", "YOUROWNNUCLEOTIDESEQUENCE")
nt_seq <- DNAStringSet(nt_seq)
msa <- msa(nt_seq, type = "dna")
print(nt_seq)
print(msa)
```

```
aa_seq =  c("ARNRRTDCDCLTQCCTQTRRBWWSRDC", "YOUROWNAMINOACIDSSEQUENCE", "ARDCMDGARNDCLTQPDGPDCQK", "TQTRRRRTDCDCGPSWBWWBWTW")
aa_seq = AAStringSet(aa_seq) 
msa <- msa(aa_seq, type = "protein")
print(aa_seq)
print(msa)
```


---


There is not a single way of attaining an MSSA and this tutorial we will use the most popular tools, including:


- Muscle [(Edgar 2004)](https://doi.org/10.1093/nar/gkh340)

- Clustal [(Higgins and Sharp 1988)](https://en.wikipedia.org/wiki/Clustal)

- T-Coffee [(Notredame et al. 2000)](https://doi.org/10.1006/jmbi.2000.4042)


And subsequently filter the alignments using a custom approach based on branch lengths!


---


As you can see, in the [fasta format](https://en.wikipedia.org/wiki/FASTA_format) each record consists of an ID and a sequence, 
of which the ID is always on a single line that starts with the ">" symbol, followed by lines containing the sequence. 
The sequences are not aligned yet: this is the reason why they contain no gaps and differ in length.
The use of short and simple IDs is strongly recommended because many programs or scripts may not work if you use spaces or hyphens.


At this stage you should have assembled your dataset into a multi-fasta file, which should look like [this](https://raw.githubusercontent.com/for-giobbe/Rphy/main/sequences/COX1.nt.fa). 
It's very important that each one of you has it's own dastaset as this may provide natural examples of problems in analyses, 
but If you still do not have a dataset use [the one I am using](https://github.com/for-giobbe/Rphy/blob/main/sequences/COX1.nt.fa).


Let's start by importing the sequences in R:


```
nt_seq <- readDNAStringSet("sequences/COX1.nt.fa")
str(nt_seq)
```


---


## aligning with Muscle:

Muscle is a very fast and versatile tool. Let's start by typing:

```
alignment_muscle <- msa(nt_seq, "Muscle")
checkAlignment(alignment_muscle2, plot = F)
```

Muscle contains many parameters, which you are encouraged to explore: important ones are the gap-opening and gap-extension penalty which by default are 400 and 0. 
Let's try a different value, using the code:

```
alignment_muscle1 <- msa(nt_seq, "Muscle", gapOpening=800, gapExtension=0)
alignment_muscle1 <- as.DNAbin(alignment_muscle1)
alignment_muscle2 <- msa(nt_seq, "Muscle", gapOpening=000, gapExtension=0)
alignment_muscle2 <- as.DNAbin(alignment_muscle2)
```

We can then inspect the outcome there is any difference between the two gap-opening penalty values and using this code:

```
checkAlignment(alignment_muscle1, plot = F)
checkAlignment(alignment_muscle2, plot = F)
```

Instead for a visual representation:

```
par(mfcol = c(2, 1))
image(alignment_muscle1, show.labels = F)
image(alignment_muscle2, show.labels = F)
```

![alt text](https://github.com/for-giobbe/Rphy/blob/main/figures/Figure_1.1.png)


## aligning with Clustal Omega:

Clustal Omega leverages [HMM profiles](https://www.ebi.ac.uk/training/online/courses/pfam-creating-protein-families/what-are-profile-hidden-markov-models-hmms/) and can be run using:

```
alignment_clustal <- msa(sequences, "ClustalOmega")
```

Another approach which I find interesting to visualise alignements is this:


```
alignment_clustal <- msa(nt_seq, "ClustalOmega")
alignment_clustal <- as.DNAbin(alignment_clustal)
par(mfcol = c(1, 3))
checkAlignment(alignment_clustal, what = 2:4)
checkAlignment(alignment_clustal, what = 3:4)
checkAlignment(alignment_clustal, what = 4:4)
```


![alt text](https://github.com/for-giobbe/Rphy/blob/main/figures/Figure_1.2.png)


Next we are moving on-line: T-Coffe is a popular aligner with an incredibly nice [server](http://tcoffee.crg.cat/) which I strongly encourage to explore. But today we are focusing on just two approaches.


---


## One aligner to rule them all: [M-coffe](https://tcoffee.crg.eu/apps/tcoffee/do:mcoffee).


The idea behind M-COFFE is to combine many of the more popular aligners - including Clustal and Muscle!
We can give it a try using the default combination of aligners (Mkalign, Muscle & MAFFT) and then take a look at the outputs, including:


- the ```.phylip``` is the alignment file in the [phylip format](https://en.wikipedia.org/wiki/PHYLIP).
- the ```.dnd``` file is the guide tree generated in the process.


As you can see M-Coffee is combining multiple aligners into one and evaluating the uncertainty between their outcome. 


---


## The tradeoff between speed and accuracy: [PSI-Coffe](https://tcoffee.crg.eu/apps/tcoffee/do:psicoffee).


The more accurate method (IMHO) is to rely on [homology extension](https://doi.org/10.1093/nar/gkw300). This approach is quite useful when dealing with distantly related proteins, 
and can be a game-changer when accurate alignments are necessary, for example for inferences on selection regimes.
PSI-Coffe works only with amino acids, so we have to translate our nucleotide sequences.

```
nt_seq <- as.DNAbin(readDNAStringSet("sequences/COX1.nt.fa"))
aa_seq <- trans(nt_seq, code = 5, codonstart = 1)
write.FASTA(aa_seq, "sequences/COX1.aa.fa", header = NULL, append = FALSE)
```

Then we can proceed to upload and process our amino acids sequences, but most likely it will take quite a long time.

---


### Importing alignments.


```
alignement <- read.alignment("alignments/COX1.nt.aln.phy", format = "phylip")
alignement <- as.DNAbin(alignement)
alview(alignement, file = "", uppercase = TRUE, showpos = TRUE)

par(mfcol = c(2, 2))

for (x in c("a", "g", "c", "t"))
  image(alignement, x, "black", cex.lab = 0.5, cex.axis = 0.7)
par(mfcol = c(1, 1))
```


![alt text](https://github.com/for-giobbe/Rphy/blob/main/figures/Figure_1.3.png)


---


## alignment filtering pt. 1:


A first approach is to remove sequences of dubious homology. There are - of course - many approaches but here is the one which I like more lately.
The following code is basically retrieving the tips which generate a branch which is over 10% of the total length of a gene tree. 


```
alignement <- read.alignment("alignments/COX1+contaminants.aln.fa", format = "phylip")

dm  <- dist.ml(alignement)
tree  <- NJ(dm)

for (x in 1:length(treeNJ$tip.label)) {
  if (tree$edge.length[sapply(1:length(tree$tip.label),function(x,y) which (y==x),y=tree$edge[,2])][x] > (sum(treeNJ$edge.length) * 0.1)) {
    print(treeNJ$tip.label[x])
  }
}
```


This approach has to be fine-tuned a bit ... the 10% cutoff should be based on the number of tips. Furthermore, it relies on the assumption that contaminants are rare ... 
can you explain why?!


---


## alignment filtering:


Other approaches for filtering are possible and useful when dealing with a large number of loci. Positions, genes and entire alignment can be filtered out on several criteria, including:


- positions' entropy
- sequences' length
- alignements' occupancy


All this approaches can be extremely helpfull in removing noise and ameliorating certain characteristic of phylogenetic datasets which can affect subsequent inferences, 
such as substitution saturation & compositional heterogeneity - more on that in the following lessons!
Here are the most popular (ranked):


- [trimal](http://trimal.cgenomics.org/trimal) [(Talavera & Castresana 2007)](https://academic.oup.com/sysbio/article/56/4/564/1682121)

- [Aliscore](https://www.zfmk.de/de/forschung/forschungszentren-und-gruppen/aliscore) [(KÃ¼ck et al. 2014)](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-15-294)

- [BMGE](https://research.pasteur.fr/en/software/bmge-block-mapping-and-gathering-with-entropy/) [(Criscuolo & Gibaldo)](https://bmcevolbiol.biomedcentral.com/articles/10.1186/1471-2148-10-210)

- [Gblocks](http://phylogeny.lirmm.fr/phylo_cgi/one_task.cgi?task_type=gblocks) [(Talavera & Castresana)](https://doi.org/10.1080/10635150701472164)

- ...


Consider also that a very stringent filtering can be a terrible approach in certain scenarios, for example when the phylogenetic signal and/or the amount of positions considered are scarce.


---


## conclusions:


In the end, you should have an aligned & filtered MSA, named something like [this](https://github.com/for-giobbe/Rphy/blob/main/alignments/COX1.nt.aln.phy).
An underlying theme of bioinformatics should start to become clear: every file should have a name which should convey most, if not all, the information regarding it. Or you will regret it ... :-P


---


## further reading: 


[very interesting paper on multiple sequence alignment uncertainty pt.1 ](https://academic.oup.com/gbe/article/7/8/2102/556628).

[very interesting paper on multiple sequence alignment uncertainty pt.2](https://science.sciencemag.org/content/319/5862/473?casa_token=t07ptffISm4AAAAA:j5l4US_y_GHOMduYw6R-MhyM7YUpa__08lw45l455DAU3tGFNKYlV40ZH0Si5w48Xl1gTEqsocLVvaE).

[filtering too heavily can worsen phylogenetic inferences](https://doi.org/10.1093/sysbio/syv033)

[alignment can substantially impact inferences of positive selection](https://doi.org/10.1093/molbev/msr272)