# Inferring Distance-Based Phylogenies


---


### intro: 


Before starting this practical set the appropriate working directory and load the necessary libraries in R:


Libraries:
- [Biostrings](https://kasperdanielhansen.github.io/genbioconductor/html/Biostrings.html#overview)
- [phangorn](https://klausvigo.github.io/phangorn/)
- [msa](https://bioconductor.org/packages/devel/bioc/vignettes/msa/inst/doc/msa.pdf)
- [ape](https://emmanuelparadis.github.io/)
- [seqinr](https://www.rdocumentation.org/packages/seqinr/versions/4.2-36)


```
setwd("~/.../RPHY/")
library(Biostrings) 
library(msa)
library(ape)
library(seqinr)
library(phangorn)
```


Before starting you should have an aligned & filtered MSA, 
or you can grab the ones I generated, 
[ND2.nt.fa](https://github.com/for-giobbe/Rphy/blob/main/sequences/ND2.nt.fa) and 
[CO1.nt.fa](https://github.com/for-giobbe/Rphy/blob/main/sequences/CO1.nt.fa).


Phylogeny reconstruction methods fall into two categories: 

- distance based  
- character based


Distance methods involve calculating a genetic distance between every pair of species (based on comparison of their aligned sequences) 
and using the resulting distance matrix iteratively to construct a tree. Character-based methods include maximum parsimony, 
maximum likelihood (ML) and Bayesian inference (BI)


---


### distance-based phylogenies: 


Let's start by reading the nucleotide sequences and aligning them as we learned during the last lesson: 

```
sequences <- readDNAStringSet("sequences/COX1.nt.fa")
alignment <- as.phyDat(msa(sequences, "Muscle"))
```


Then we proceed to calculate a distance matrix based on p distances.
This distance is the proportion (p) of nucleotide sites at which two sequences being compared are different. 
It is obtained by dividing the number of nucleotide differences by the total number of nucleotides compared. 
It does not make any correction for multiple substitutions at the same site, substitution rate biases 
(for example, differences in the transitional and transversional rates), or differences in evolutionary rates among sites.


```
dm  <- dist.p(alignment)
```


After constructing the distance matrix, we infer:


- a rooted tree with [unweighted pair group method with arithmetic mean](https://en.wikipedia.org/wiki/UPGMA)
 
 
```
treeUPGMA  <- upgma(dm)
```


-  an unrooted tree using [Neighbor Joining](https://en.wikipedia.org/wiki/Neighbor_joining).


```
treeNJ  <- nj(dm)
```


Then we can visualize differences using a so-called cophylo plot:


```
treeNJ$edge.length<-NULL
treeUPGMA$edge.length<-NULL
cop<-cophylo(treeUPGMA,treeNJ)
plot(cop,link.type="curved",link.lwd=5,link.lty="solid", link.col=make.transparent("blue",0.25),fsize=0.5)
```


![alt text](https://github.com/for-giobbe/Rphy/blob/main/figures/Figure_2.1.png)




---


### parsimony:


parsimony(treeUPGMA, alignment)
parsimony(treeNJ, alignment)



---


### the newick format:

The Newick is by far the most used format to store trees and it substantially is a combinations of parentheses, punctuation, numbers & letters.
The format name is known to have a [quite funny origin](http://newicks.com/). Lets take a look at some possibilities the ```.nwk``` offers:


```
(,,(,));                         	      no nodes are named

(A,B,(C,D));                     	      leaf nodes are named

(A,B,(C,D)E)F;                     	      all nodes are named

(:0.1,:0.2,(:0.3,:0.4):0.5);    	      all but root node have a distance to parent

(:0.1,:0.2,(:0.3,:0.4):0.5):0.0;     	      all have a distance to parent

(A:0.1,B:0.2,(C:0.3,D:0.4):0.5);    	      distances and leaf names (popular)

(A:0.1,B:0.2,(C:0.3,D:0.4)E:0.5)F; 	      distances and all names

((B:0.2,(C:0.3,D:0.4)E:0.5)A:0.1)F; 	      a tree rooted on a leaf node (rare)
```

Such a simple format has several interesting implications. If N = n. species, there are:

* N terminal branches
* 2N-3 total branches
* N-3 internal branches
* N-2 internal nodes

Regarding the number of trees which can possibly describe the relationships between a given number of terminal nodes (_e.g._ species),
I'll just say that is a number which goes up quite quickly. This even more true for trees with a root, as 
for each unrooted there are 2N-3 times as many rooted trees. Here are some numbers: 

| Leaves        | Unrooted trees           | Rooted trees  |
| ------------- |:------------------------:| -------------:|
| 3             | 1                        | 3             |
| 5             | 15                       | 105           |
| 7             | 945                      | 10,395        |
| 9             | 135,135                  | 2,027,025     |
| 10            | 2,027,025                | 34,459,425    |

Last but not least, remember that ```.nwk``` trees are not unique representations, 
and that relationships between terminals can be written in several different ways.

We can visualize ```.nwk``` files either [online](https://itol.embl.de/upload.cgi) or using software as [FigTree](http://tree.bio.ed.ac.uk/software/figtree/).


---


### further reading: 


[Phylogenetic tree building in the genomic age](https://www.nature.com/articles/s41576-020-0233-0).

[FastME2.0](https://doi.org/10.1093/molbev/msv150)

[APPLES](https://doi.org/10.1093/sysbio/syz063)