# Inferring Distance-Based Phylogenies


---


### intro: 


Before starting this practical set the appropriate working directory and load the necessary libraries in R:


Libraries:
- [Biostrings](https://kasperdanielhansen.github.io/genbioconductor/html/Biostrings.html#overview)
- [phangorn](https://klausvigo.github.io/phangorn/)
- [msa](https://bioconductor.org/packages/devel/bioc/vignettes/msa/inst/doc/msa.pdf)
- [ape](https://emmanuelparadis.github.io/)
- [seqinr](https://www.rdocumentation.org/packages/seqinr/versions/4.2-36)


```
setwd("~/.../RPHY/")
library(Biostrings) 
library(msa)
library(ape)
library(seqinr)
library(phangorn)
```


Before starting you should have an aligned & filtered MSA, 
or you can grab the one I am using: 
[CO1.nt.fa](https://github.com/for-giobbe/Rphy/blob/main/sequences/CO1.nt.fa).


Phylogeny reconstruction methods fall into two categories: 

- distance based  
- character based


Distance methods involve calculating a genetic distance between every pair of species (based on comparison of their aligned sequences) 
and using the resulting distance matrix iteratively to construct a tree. Character-based methods include maximum parsimony, 
maximum likelihood (ML) and Bayesian inference (BI)


---


### distance-based phylogenies: 


Let's start by reading the nucleotide sequences and aligning them as we learned during the last lesson: 

```
sequences <- readDNAStringSet("sequences/COX1.nt.fa")
alignment <- as.phyDat(msa(sequences, "Muscle"))
```


Then we proceed to calculate a distance matrix based on p distances.
This distance is the proportion (p) of nucleotide sites at which two sequences being compared are different. 
It is obtained by dividing the number of nucleotide differences by the total number of nucleotides compared. 
It does not make any correction for multiple substitutions at the same site, substitution rate biases 
(for example, differences in the transitional and transversional rates), or differences in evolutionary rates among sites.


```
dm  <- dist.p(alignment)
```


After constructing the distance matrix, we infer:


- a rooted tree with [Unweighted Pair Group Method with Arithmetic mean (UPGMA)](https://en.wikipedia.org/wiki/UPGMA)
 
 
```
treeUPGMA  <- upgma(dm)
```


-  an unrooted tree using [Neighbor Joining (NJ)](https://en.wikipedia.org/wiki/Neighbor_joining)


```
treeNJ  <- nj(dm)
```


Then we can visualize differences using a so-called cophylo plot:


```
treeNJ$edge.length<-NULL
treeUPGMA$edge.length<-NULL
cop<-cophylo(treeUPGMA,treeNJ)
plot(cop,link.type="curved",link.lwd=5,link.lty="solid", 
link.col=make.transparent("blue",0.25),fsize=0.5)
```


![alt text](https://github.com/for-giobbe/Rphy/blob/main/figures/Figure_2.1.png)


---

### nodal support:

Phylogenetic trees are almost always just hypotheses and they should be treated as such! 
In this perspective, it's important to know how much we can be confident in respect to our findings: 
there are several possibilities which take into consideration the whole tree, 
but it's also possible to calculating several metrics of clade support, specifically for each node. 
The more frequently used metric by far is the bootstrap.


I really like this explanation of parametric and non-parametric bootstrap:

> Non-parametric bootstrapping was developed by Efron in 1979 as a general statistical method for estimating the parameters 
> of an unknown probability distribution by resampling from an existing sample that was drawn from this distribution. 
> The method was transferred to phylogenetic reconstruction by Felsenstein in [this](https://onlinelibrary.wiley.com/doi/abs/10.1111/j.1558-5646.1985.tb00420.x) paper from 1985.
> Within molecular phylogenetics it works as follows: 
> from an alignment of length n, columns are randomly drawn with replacement n times. 
> The drawn columns are arranged in a new dataset, a bootstrapped alignment of length n. 
> From this bootstrapped alignment, a phylogenetic tree is constructed by following the same method of phylogenetic analysis 
> as was used for the analysis of the original alignment. 
> This process of constructing bootstrap alignments and bootstrap trees is repeated a large number of times,
> and the resulting trees are stored. The percentage with which a certain bipartition of the taxon set is present in the bootstrap trees
> (the bootstrap value) can be taken as a measure of how homogeneously this bipartition of sequences 
> (i.e., the respective branch in the underlying topology) is supported by the data. 
> Bootstrap values are often summarized by constructing the majority-rule consensus from the bootstrap trees
> or by annotating them on the "best" tree.

Along with its non-replacement version (not too popular nowadays):

> The jackknife is a resampling method closely related to non-parametric bootstrapping. 
> It works by randomly deleting a certain percentage of columns from the original alignment. 
> Usually 50 per cent of the columns are deleted (delete-half jackknife proposed by Felsenstein in 1985), 
> which is equivalent to drawing n/2 columns from the original alignment of length n without replacement. 
> As in non-parametric bootstrapping, the resampling is iterated, trees are computed from the jackknife alignments, 
> and branch-support values are derived as the percentage with which a certain branch is present in the jackknife topologies.


Anyway, to obtain the bootstrap trees use:


```
bstrees <- bootstrap.phyDat(alignment,
                            FUN=function(alignment)NJ(dist.p(alignment)), 
                            bs=1000)
```


Subsequently, you can take a look at some bootstrap trees and then
summarise them on your original phylogeny:


```
plot(bstrees)
plotBS(treeNJ, bstrees, type = "phylogram")
```


---


### parsimony:


To infer phylogenies with MP (maximum parsimony) we will use the parsimony ratchet (Nixon 1999). 
This allows to escape local optima and find better trees than only performing NNI / SPR rearrangements.

The current implementation is

- Create a bootstrap data set Db from the original data set.
- Take the current best tree and perform tree rearrangements on Db and save bootstrap tree as Tb
- Use Tb and perform tree rearrangements on the original data set. If this tree has a lower parsimony score than the currently best tree, replace it.
- Iterate until either a given number of iteration is reached (minit) or no improvements have been recorded for a number of iterations (k).


```
treePARSIMONY  <- pratchet(alignment,
                            maxit=200, minit=50, 
                            k=50, all = T)
```

```
fitch(treeUPGMA, alignment)
fitch(treeNJ, alignment)
fitch(treePARSIMONY, alignment)
```

```
treePARSIMONY1 <- treePARSIMONY[[1]]
treePARSIMONY1 <- root(treePARSIMONY1,
which((treePARSIMONY1$tip.label == "Mantis_religiosa") == "TRUE"))

treePARSIMONY2 <- treePARSIMONY[[2]]
treePARSIMONY2 <- root(treePARSIMONY2,
which((treePARSIMONY2$tip.label == "Mantis_religiosa") == "TRUE"))

cop <- cophylo(treePARSIMONY1,treePARSIMONY2)
plot(cop,link.type="curved",link.lwd=5,link.lty="solid", link.col=make.transparent("blue",0.25),fsize=0.5)
```


---


### the newick format:

The Newick is by far the most used format to store trees and it substantially is a combinations of parentheses, punctuation, numbers & letters.
The format name is known to have a [quite funny origin](http://newicks.com/). Lets take a look at some possibilities the ```.nwk``` offers:


```
(,,(,));                         	      no nodes are named

(A,B,(C,D));                     	      leaf nodes are named

(A,B,(C,D)E)F;                     	      all nodes are named

(:0.1,:0.2,(:0.3,:0.4):0.5);    	      all but root node have a distance to parent

(:0.1,:0.2,(:0.3,:0.4):0.5):0.0;     	      all have a distance to parent

(A:0.1,B:0.2,(C:0.3,D:0.4):0.5);    	      distances and leaf names (popular)

(A:0.1,B:0.2,(C:0.3,D:0.4)E:0.5)F; 	      distances and all names

((B:0.2,(C:0.3,D:0.4)E:0.5)A:0.1)F; 	      a tree rooted on a leaf node (rare)
```

Such a simple format has several interesting implications. If N = n. species, there are:

* N terminal branches
* 2N-3 total branches
* N-3 internal branches
* N-2 internal nodes

Regarding the number of trees which can possibly describe the relationships between a given number of terminal nodes (_e.g._ species),
I'll just say that is a number which goes up quite quickly. This even more true for trees with a root, as 
for each unrooted there are 2N-3 times as many rooted trees. Here are some numbers: 

| Leaves        | Unrooted trees           | Rooted trees  |
| ------------- |:------------------------:| -------------:|
| 3             | 1                        | 3             |
| 5             | 15                       | 105           |
| 7             | 945                      | 10,395        |
| 9             | 135,135                  | 2,027,025     |
| 10            | 2,027,025                | 34,459,425    |

Last but not least, remember that ```.nwk``` trees are not unique representations, 
and that relationships between terminals can be written in several different ways.

We can visualize ```.nwk``` files either [online](https://itol.embl.de/upload.cgi) or using software as [FigTree](http://tree.bio.ed.ac.uk/software/figtree/).


---


### further reading: 


[Phylogenetic tree building in the genomic age](https://www.nature.com/articles/s41576-020-0233-0).

[FastME2.0](https://doi.org/10.1093/molbev/msv150)

[APPLES](https://doi.org/10.1093/sysbio/syz063)