### Inferring ML Phylogenies


---


### intro: 


Before starting this practical set the appropriate working directory and load the necessary libraries in R:


Libraries:


- [Biostrings](https://kasperdanielhansen.github.io/genbioconductor/html/Biostrings.html#overview)
- [phangorn](https://klausvigo.github.io/phangorn/)
- [msa](https://bioconductor.org/packages/devel/bioc/vignettes/msa/inst/doc/msa.pdf)
- [ape](https://emmanuelparadis.github.io/)
- [seqinr](https://www.rdocumentation.org/packages/seqinr/versions/4.2-36)
- [phytools](http://www.phytools.org/)


```
setwd("~/Desktop/RPHY/")
library(Biostrings) 
library(msa)
library(ape)
library(seqinr)
library(phangorn)
library(phytools)
```


You should have 2 ```fasta``` files, 
or you can use these: 


- [CO1.nt.fa](https://github.com/for-giobbe/Rphy/blob/main/sequences/CO1.nt.fa)
- [ND2.nt.fa](https://github.com/for-giobbe/Rphy/blob/main/sequences/ND2.nt.fa)
- [ND3.nt.fa](https://github.com/for-giobbe/Rphy/blob/main/sequences/ND3.nt.fa)


---


Maximum likelihood is a general statistical method for estimating parameters of a
probability model: for example a normal distribution can be described by two parameters: 
the mean and variance. Instead, in molecular phylogenetics there are a wide plethora of parameters, which include:


* rates of transitions / trasversions / ... between bases
* base composition
* descriptors of rate heterogeneity across sites
* branchlengths
* the tree itself!


The likelihood is defined as a quantity proportional to the probability of observing the data given the model.
From an operational point of view it consists in finding 
the tree topology that maximizes the probability of observing the given 
sequence data under a specific model of evolution. 


```
P(D|M)
```


If we have a model, we can calculate the probability the observations would have actually been observed as a function of the model. 
We then examine this likelihood function to see where it is at its greatest, and the value of the parameter of
interests at that point is the maximum likelihood estimate of the parameter (_e.g._ the tree and branch lengths).


---


### inferring ML gene trees - unpartitioned analyses:


Let's start by inferring a NJ tree for a single genes, just
as we learned during the last lesson: 


```
sequences <- readDNAStringSet("sequences/COX1.nt.fa")
alignment <- as.phyDat(msa(sequences, "Muscle"))
nj_tree <- nj(dist.ml(alignment))
```


Then we conduct a model test to determine the best-fitting evolutionary model, based on AIC:


```
mt = modelTest(alignment, nj_tree)
var_mt = attr(mt, "env")
evolutionary_model = eval(get(mt$Model[which.min(mt$AIC)], var_mt), var_mt)
```

Which gives us a list of models and their fit.


```
     Model df     logLik      AIC          AICw     AICc         AICcw      BIC
1       JC 33 -10757.382 21580.76  0.000000e+00 21583.03  0.000000e+00 21743.50
2     JC+I 34 -10439.512 20947.02  0.000000e+00 20949.43  0.000000e+00 21114.69
3     JC+G 34 -10341.964 20751.93  0.000000e+00 20754.33  0.000000e+00 20919.60
4   JC+G+I 35 -10337.454 20744.91  0.000000e+00 20747.46  0.000000e+00 20917.51
5      F81 36 -10295.979 20663.96  0.000000e+00 20666.66  0.000000e+00 20841.49
6    F81+I 37  -9910.150 19894.30 6.571190e-194 19897.15 1.088452e-193 20076.76
7    F81+G 37  -9758.870 19591.74 3.295440e-128 19594.59 5.458568e-128 19774.20
8  F81+G+I 38  -9756.606 19589.21 1.165573e-127 19592.22 1.784708e-127 19776.61
9      K80 34 -10750.365 21568.73  0.000000e+00 21571.14  0.000000e+00 21736.40
10   K80+I 35 -10430.979 20931.96  0.000000e+00 20934.51  0.000000e+00 21104.56
11   K80+G 35 -10330.350 20730.70  0.000000e+00 20733.25  0.000000e+00 20903.30
12 K80+G+I 36 -10326.032 20724.06  0.000000e+00 20726.76  0.000000e+00 20901.60
13     HKY 37 -10193.844 20461.69 4.081387e-317 20464.54 6.760411e-317 20644.15
14   HKY+I 38  -9730.198 19536.40 3.431665e-116 19539.41 5.254513e-116 19723.79
15   HKY+G 38  -9498.041 19072.08  2.292155e-15 19075.09  3.509712e-15 19259.48
16 HKY+G+I 39  -9489.896 19057.79  2.904617e-12 19060.96  4.102292e-12 19250.12
17     SYM 38 -10340.972 20757.94  0.000000e+00 20760.95  0.000000e+00 20945.34
18   SYM+I 39 -10019.349 20116.70 3.345708e-242 20119.87 4.725261e-242 20309.03
19   SYM+G 39  -9922.043 19922.09 6.079005e-200 19925.26 8.585591e-200 20114.41
20 SYM+G+I 40  -9913.167 19906.33 1.601945e-196 19909.67 2.082286e-196 20103.59
21     GTR 41 -10132.566 20347.13 3.064546e-292 20350.64 3.658106e-292 20549.32
22   GTR+I 42  -9696.127 19476.25 3.936451e-103 19479.94 4.305577e-103 19683.38
23   GTR+G 42  -9471.059 19026.12  2.193034e-05 19029.80  2.398678e-05 19233.24
24 GTR+G+I 43  -9459.331 19004.66  9.999781e-01 19008.52  9.999760e-01 19216.72
```


After this practical, models of sequence evolutions shouldn't be 
something obscure ... they are just a handfull of numbers in the end!


```
 loglikelihood: -9424.35 

unconstrained loglikelihood: -5600.794 
Proportion of invariant sites: 0.2121178 
Discrete gamma model
Number of rate categories: 4 
Shape parameter: 0.7337186 

Rate matrix:
          a         c        g         t
a 0.0000000 0.8122448 2.785374 0.6710854
c 0.8122448 0.0000000 1.849860 7.8686555
g 2.7853745 1.8498604 0.000000 1.0000000
t 0.6710854 7.8686555 1.000000 0.0000000

Base frequencies:  
0.466208 0.09923958 0.05858567 0.3759667 
```


![Alt text](https://lh6.googleusercontent.com/proxy/dBql_qGCbb3Q8ozYu9FWYxq-VP7epgRMgQLagE_YOEax2h5omyzB-o8DBO5-KkmztPXx4xrRJnBvaU198uPuitHr1wmHK4jrz6P_jqvmyJlMg6IfzjrWMlJw "MK model")


Then we can﻿compute the likelihood of a phylogenetic tree 
given a sequence alignment and a model, specifying the following parameters:


- __bf__ for base frequencies
- __Q__ for the rate matrix
- __inv__ for the proportion of invariant sites
- __k__ for the number of rate categories
- __shape__ for the gamma distribution shape parameter


```
initial_pml = pml(tree = nj_tree, data = alignment, 
                  model = evolutionary_model$model,
                  bf = evolutionary_model$bf, 
                  Q = evolutionary_model$Q, 
                  inv = evolutionary_model$inv,
                  k = evolutionary_model$k, 
                  shape = evolutionary_model$shape)
```


Which gives us a ```pml``` object with its parameters.


```
 loglikelihood: -9750.387 

unconstrained loglikelihood: -5600.794 
Proportion of invariant sites: 0.2121178 
Discrete gamma model
Number of rate categories: 4 
Shape parameter: 0.7337186 

Rate matrix:
          a         c        g         t
a 0.0000000 0.8122448 2.785374 0.6710854
c 0.8122448 0.0000000 1.849860 7.8686555
g 2.7853745 1.8498604 0.000000 1.0000000
t 0.6710854 7.8686555 1.000000 0.0000000

Base frequencies:  
0.466208 0.09923958 0.05858567 0.3759667 
```

Then we can further optimise the parameters, based on the tree:

```
pml = optim.pml(initial_pml, 
				model = evolutionary_model$model, 
            	optInv = T, 
                optGamma = F,
                optBf = F, 
                optQ = F,
                optEdge = F,
                optNni = TRUE,
                rearrangement = "NNI")
```


Which gives us a ```pml``` object with different parameters.


```
 loglikelihood: -9424.127 

unconstrained loglikelihood: -5600.794 
Proportion of invariant sites: 0.2081278 
Discrete gamma model
Number of rate categories: 4 
Shape parameter: 0.7337186 

Rate matrix:
          a         c        g         t
a 0.0000000 0.8122448 2.785374 0.6710854
c 0.8122448 0.0000000 1.849860 7.8686555
g 2.7853745 1.8498604 0.000000 1.0000000
t 0.6710854 7.8686555 1.000000 0.0000000

Base frequencies:  
0.466208 0.09923958 0.05858567 0.3759667 
```


Then we can plot the trees of:

- the NJ algorithm
- the initial ```pml``` object, tha same as the NJ one
- the ML optimised ```pml``` object,

```
par(mfrow = c(1, 3))
plot(nj_tree, use.edge.length = F)
plot(initial_pml$tree, use.edge.length = F)
plot(pml$tree, use.edge.length = F)
```


![Alt text](https://github.com/for-giobbe/Rphy/blob/main/figures/Figure_3.1.png)


---


### Additional node support metrics - TBE:


The difference between the Transfer Bootstrap Expectation (TBE)
and the non-parametric bootstrap proportions(BP) 
lies in the comparison of the pseudo-trees to the reference tree. 
Rather than the binary presence/absence of a reference branch in the pseudo-trees, 
TBE uses a “transfer” distance that is measured using the number of taxa that must 
be transferred (or removed) to make two branches identical.


```
bs <- bootstrap.pml(pml, bs=100, optNni=TRUE)
plotBS(pml$tree, bs, "p", method = "TBE")
```


---


### inferring ML species trees - partitioned analyses:


Let's import, align and infer an evolutionary model for three genes:


```
sequences_cox1 <- readDNAStringSet("sequences/COX1.nt.fa")
alignment_cox1 <- as.phyDat(msa(sequences_cox1))
nj_tree_cox1 <- nj(dist.ml(alignment_cox1))
mt_cox1 = modelTest(alignment_cox1, nj_tree_cox1)
var_mt_cox1 = attr(mt_cox1, "env")
evolutionary_model_cox1 = eval(get(mt_cox1$Model[which.min(mt_cox1$AIC)], 
							   var_mt_cox1), var_mt_cox1)
pml_cox1 <- optim.pml(pml(tree = nj_tree_cox1, data = alignment_cox1,
                          model = evolutionary_model_cox1$model,
                          bf = evolutionary_model_cox1$bf, 
                          Q = evolutionary_model_cox1$Q, 
                          inv = evolutionary_model_cox1$inv,
                          k = evolutionary_model_cox1$k, 
                          shape = evolutionary_model_cox1$shape))

sequences_nad2 <- readDNAStringSet("sequences/NAD2.nt.fa")
alignment_nad2 <- as.phyDat(msa(sequences_nad2))
nj_tree_nad2 <- nj(dist.ml(alignment_nad2))
mt_nad2 = modelTest(alignment_nad2, nj_tree_nad2)
var_mt_nad2 = attr(mt_nad2, "env")
evolutionary_model_nad2 = eval(get(mt_nad2$Model[which.min(mt_nad2$AIC)], 
							   var_mt_nad2), var_mt_nad2)
pml_nad2 <- optim.pml(pml(tree = nj_tree_nad2, data = alignment_nad2,
                          model = evolutionary_model_nad2$model,
                          bf = evolutionary_model_nad2$bf, 
                          Q = evolutionary_model_nad2$Q, 
                          inv = evolutionary_model_nad2$inv,
                          k = evolutionary_model_nad2$k, 
                          shape = evolutionary_model_nad2$shape))
                          
sequences_nad3 <- readDNAStringSet("sequences/NAD3.nt.fa")
alignment_nad3 <- as.phyDat(msa(sequences_nad3))
nj_tree_nad3 <- nj(dist.ml(alignment_nad3))
mt_nad3 = modelTest(alignment_nad3, nj_tree_nad2)
var_mt_nad3 = attr(mt_nad3, "env")
evolutionary_model_nad3 = eval(get(mt_nad3$Model[which.min(mt_nad3$AIC)], 
                                   var_mt_nad3), var_mt_nad3)
pml_nad3 <- optim.pml(pml(tree = nj_tree_nad3, data = alignment_nad3,
                          model = evolutionary_model_nad3$model,
                          bf = evolutionary_model_nad3$bf, 
                          Q = evolutionary_model_nad3$Q, 
                          inv = evolutionary_model_nad3$inv,
                          k = evolutionary_model_nad3$k, 
                          shape = evolutionary_model_nad3$shape))
```


Then we can jointly optimize parameters of multiple partitions, using formulas such as:
```parameter1 ~ parameter2 + parameter3```. 
On the left side of ```~``` are the parameters which get optimized over all partitions, 
on the right of ```~``` are the parameter which are optimized specifically to each partition. 
Here we optimize:


- the base frequencies jointy for all partition
- the tree separately for each partition


```
pr1 <- pmlPart(bf ~ nni, object = list(pml_cox1, pml_nad2, pml_nad3), optNni = TRUE)
```


... and we can see that inside the ```pml``` object there are three different trees:


```
par(mfrow = c(1, 3))
plot.phylo(pr1$fits[[1]]$tree)
plot.phylo(pr1$fits[[2]]$tree)
plot.phylo(pr1$fits[[3]]$tree)
```


Then we optimize:


- the tree jointy for all partition
- the proportion of invariant sites separately for each partition

```
pr2 <- pmlPart(nni ~ inv, object = list(pml_cox1, pml_nad2, pml_nad3), optNni = TRUE)
```


... and we can see that inside the ```pml``` object there are three identical trees:


```
par(mfrow = c(1, 3))
plot.phylo(pr2$fits[[1]]$tree)
plot.phylo(pr2$fits[[2]]$tree)
plot.phylo(pr2$fits[[3]]$tree)
```


Then we can use the SH (Shimodaira-Hasegawa) test is a resampling method that uses site likelihood scores 
to test the null hypothesis that "tree fit to data is same".
Rejection of the null hypothesis is justification for not combining the partitions.


```
SH.test(pml_cox1, pml_nad2, pml_nad3, B=10000)
```


... what do you think we should do!?


---


### conclusions: 


In this lesson we:


- carried out a ML phylogenetic inference on both single loci and partitions
- explored metrics of nodal support often associated to ML (BP and TBE)
- learned how to scale-up things from a single gene-tree to multiple-loci


---


### further reading: 


Some [music](https://youtu.be/pZ12_E5R3qc?t=26) which has been clearly inspired by phylogenetic methods

Nice [paper](https://academic.oup.com/sysbio/article/72/6/1280/7240218) on the robustness of BP versus TBE


---


[0](https://github.com/for-giobbe/Rphy/blob/main/markdowns/0.md)
[1](https://github.com/for-giobbe/Rphy/blob/main/markdowns/1.md)
[2](https://github.com/for-giobbe/Rphy/blob/main/markdowns/2.md)
[3](https://github.com/for-giobbe/Rphy/blob/main/markdowns/3.md)
[4](https://github.com/for-giobbe/Rphy/blob/main/markdowns/4.md)
[5](https://github.com/for-giobbe/Rphy/blob/main/markdowns/5.md)